File: A2Z_striver/patterns/pattern_15/pattern_15.py
import sys

class Solution:
    """
    Problem: Print Inverted Increasing Letter Triangle.
    Example Input: n = 5
    Example Output:
    A B C D E
    A B C D
    A B C
    A B
    A
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Nested Loops (Stateless)
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern15_stateless(self, n):
        for i in range(n):
            # // - Inner loop shrinks as i increases
            for j in range(n - i):
                # // - Calculate char based on COLUMN (j)
                print(chr(ord('A') + j), end=" ")
            print()

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    n = 5
    print(f"Test Case (n={n}):")
    solver.pattern15_stateless(n)

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_15/pattern_15.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print Inverted Increasing Letter Triangle.
// Example Input: n = 5
// Example Output:
// A B C D E
// A B C D
// A B C
// A B
// A
// Goal: O(N^2) Time.

class Solution
{
public:
    // ~ Approach: Stateless (ASCII Math)
    // Analysis: Time O(N^2) | Space O(1)
    void pattern15(int n)
    {
        for (int i = 0; i < n; i++)
        {

            // - Logic: Loop runs (n-i) times
            // - Row 0: n times
            // - Row 1: n-1 times
            for (int j = 0; j < n - i; j++)
            {

                // - Calculate char based on COLUMN (j)
                // - j resets to 0 every row, so every row starts with 'A'
                char ch = 'A' + j;
                cout << ch << " ";
            }
            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 5;

    cout << "Test Case (n=" << n << "):\n";
    sol.pattern15(n);

    return 0;
}